src/status_im2/subs/chat/messages.cljs

```diff
@@ -102,6 +102,25 @@
  (fn [messages]
    (empty? messages)))
 
+(defn find-albums
+  [messages]
+  (let [albums (atom {})]
+    (reduce (fn [acc message]
+              (let [album-id (when (:albumize? message) (:album-id message))]
+                (if (and album-id (get @albums album-id))
+                  (swap! albums update-in [album-id] #(conj % message))
+                  (swap! albums assoc-in [album-id] [message]))
+                (conj acc message)
+                (if (and album-id (> (count (get @albums album-id)) 3))
+                  (conj (filterv #(not= album-id (:album-id %)) acc)
+                        {:album        (get @albums album-id)
+                         :album-id     album-id
+                         :message-id   album-id
+                         :content-type constants/content-type-album})
+                  (conj acc message))))
+            []
+            messages)))
+
```

Comment:
@ilmotta @flexsurfer does it look better now?

Reply:
yeah i was looking at this, sorry for the delay, i'll take a look on your solution, for now it seems like `(conj acc message)` in the middle is a leftover

Reply:
```clojure
(get (reduce (fn [{:keys [messages albums]} message]
               (let [album-id (when (:albumize? message) (:album-id message))
                     albums (cond-> albums album-id (update album-id #(conj % message)))
                     messages (if (and album-id (> (count (get albums album-id)) 3))
                                (conj (filterv #(not= album-id (:album-id %)) messages)
                                      {:album        (get albums album-id)
                                       :album-id     album-id
                                       :message-id   album-id
                                       :content-type constants/content-type-album})
                                (conj messages message))]
                   {:messages messages
                    :albums albums}))
             {:messages []
              :albums {}}
             messages)
     :messages)
```
this is my suggestion, but honestly im not sure how it works, condition looks strange

Reply:
`:albumize?` is confusing

Reply:
because we don't check it in the filter

Reply:
@OmarBasem, we really need tests for this kind of code to be merged in prod. The nice thing is that even if you change the implementation of `find-albums`, unit tests would not care if coded correctly, they would still be the same and you'd be free to refactor knowing that it all works.

Reply:
@OmarBasem, as @flexsurfer pointed out, we shouldn't use an atom to store temporary state, just use a map as the initial reducing value. Atom deref's are much more expensive than simple maps, which is particularly important in subscriptions.

Reply:
@OmarBasem, I reviewed in more detail the changed subscription `:chats/raw-chat-messages-stream`. A couple points:

1. The final call to `find-albums` seems to be a breaking change. I haven't seen this PR update all the other places where the subscription is used, and there are a handful. Are you sure everything is really working? I actually don't understand how it could work, maybe it's me being shortsighted.

Original implementation is a sequence of message entities with some additional keys. New implementation is also a vector, but when the messages have an album ID, they're grouped into a new map that's incompatible with a message entity.

Places using the affected subscription:

- `status-im.ui.screens.chat.views/messages-view`
- `status-im2.contexts.chat.messages.list.view/messages-list`
- Subscription `:chats/profile-messages-stream`.

2. I would also suggest changing the name `find-albums`. `find` is a word used for search operations, and searching doesn't generally change the shape of the data being searched.
   
3. Why do we need the `albumize?` keyword assoc'ed to messages? Wouldn't it be enough to simply check the existence of the `:album-id` key? In this PR we see the function `models.loading/mark-album` does exactly this, it always assocs `:albumize?` as true if there's a `:album-id` value, so `:albumize?` seems to be redundant.

4. A good chunk of the logic in `find-albums` is related to indexing messages by `album-id`, but this can be simplified by using `group-by`.

5. To reiterate, missing tests :man_shrugging: 

Feel free to ping me if you'd like to pair program on some of these points. I have some ideas, but writing it all takes too much time via GH threads.

Reply:
> could you please explain in words these two lines ?
> 
> ```clojure
> (if (and album-id (> (count (get albums album-id)) 3))
>               (conj (filterv #(not= album-id (:album-id %)) acc)
> ```

We group images into albums with a count of more than 3. Before adding the "album" message into the list of messages, we need to filter out the image messages of that album that was added into the list.